### PR TITLE
Boris/tim adapter

### PR DESCRIPTION
**Pull request by borisjaskerovich**
i did not test the function that interprets the response.
i need to verify also that i am making the request as i should.

these are the params that the partner will need to supply:

   params: {
                    sizes: [[300, 250]],
                    placementCode: 'header-bid-tag-0',
                    publisherid: "testpublisherid",
                    bidfloor:1,
                    ip:<ip>
                }

### CODE REVIEW

**Comment by ghost** [comment]
Build Passed

**Comment by ghost** [comment]
Build Passed

**Comment by shlomisut** [comment]
I think the `sizes` part in the params is redundant. you should have that from the prebid auction details

File: modules/timBidAdapter.js [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+  code: BIDDER_CODE,
+  aliases: ['tim'], // short code
+  /**
+   * Determines whether or not the given bid request is valid.
+   *
+   * @param {BidRequest} bid The bid params to validate.
+   * @return boolean True if this is a valid bid, and false otherwise.
+   */
+  isBidRequestValid: function(bid) {
+    if(bid.params && bid.params.publisherid && bid.params.placementCode && bid.params.bidfloor){
```
Comment (shlomisut) [inline modules/timBidAdapter.js]: bidfloor should be optional

**Comment by ghost** [comment]
Build Passed

**Comment by ghost** [comment]
Build Passed

**Comment by ghost** [comment]
Build Passed

**Comment by ghost** [comment]
Build Passed

**Comment by ghost** [comment]
Build Passed

**Comment by ghost** [comment]
Build Passed

File: modules/timBidAdapter.js [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+      device:{
+         "language": this.getLanguage(),
+         "w":adW,
+         "h":adH,
+         "js":1,
+         "ua": navigator.userAgent
+      }
+    };
+    if(!bidFloor){
+      delete bidRequest.imp["bidfloor"];
```
[the remark this answers is not in the document]
Comment (borisjaskerovich) [reply modules/timBidAdapter.js:112]: yes

File: modules/timBidAdapter.js [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+    for (var i = 0; i < validBidRequests.length; i++) {
+      requests.push(this.requestBid(validBidRequests[i]));
+    }
+    return requests;
+  },
+
+  requestBid:function(bidReq){
+    // build bid request object
+    var domain = window.location.host;
+    var page = window.location.host + window.location.pathname + location.search + location.hash;
```
[the remark this answers is not in the document]
Comment (borisjaskerovich) [reply modules/timBidAdapter.js]: yes it gives the same result.
window.location.href is with the http also.

[the remark this answers is not in the document]
Comment (shlomisut) [reply modules/timBidAdapter.js]: also - I think there's an inherit prebid method to get the page URL, best use that

File: modules/timBidAdapter.js [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+  return JSON.parse(obj.br);
+}
+
+
+function formatAdMarkup(bid) {
+  var adm = bid.adm;
+  if ('nurl' in bid) {
+    adm += utils.createTrackPixelHtml(bid.nurl);
+  }
+  return "<!DOCTYPE html><html><head><title></title><body style='margin:0px;padding:0px;'>"+adm+"</body></head>";
```
[the remark this answers is not in the document]
Comment (borisjaskerovich) [reply modules/timBidAdapter.js:40]: i am not sure. i copied this function from some other adapter.

**Comment by ghost** [comment]
Build Passed

File: modules/timBidAdapter.js [hunk 5]
```diff
@@ -42,12 +42,12 @@ function formatAdMarkup(bid) {
 
 export const spec = {
   code: BIDDER_CODE,
-  aliases: ['tim'], // short code
 
   isBidRequestValid: function(bid) {
     if(bid.params && bid.params.publisherid && bid.params.placementCode){
       return true;
     }
+    utils.logError("error bid: " + JSON.stringify(bid));
```
Comment (shlomisut) [inline modules/timBidAdapter.js]: this doesn't add enough info.
please make a clear check and log as to what param is missing

**Comment by ghost** [comment]
Build Passed

File: modules/timBidAdapter.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (shlomisut) [reply modules/timBidAdapter.js:40]: @borisjaskerovich please look at the memegen adapter - aswell as our own previous adapter 
https://github.com/MemeGlobal/Prebid.js/blob/prebid-0.34/modules/memeglobalBidAdapter.js

[the remark this answers is not in the document]
Comment (borisjaskerovich) [reply modules/timBidAdapter.js:40]: dima asked me to add  utils.createTrackPixelHtml . it does not work as it was in our previous adapter with this function.

[the remark this answers is not in the document]
Comment (shlomisut) [reply modules/timBidAdapter.js:40]: that's not related to this issue. 
you're loading our creative in the HEADER, which is not where it usually loads.

**Comment by shlomisut** [comment]
@borisjaskerovich  is this wainting on anything? everything working ?

File: modules/timBidAdapter.js [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+  buildRequests: function(validBidRequests,bidderRequest) {
+    bidsRequested=bidderRequest;
+    var requests = [];
+    for (var i = 0; i < validBidRequests.length; i++) {
+      requests.push(this.createRTBRequestUTL(validBidRequests[i]));
+    }
+    return requests;
+  },
+
+  createRTBRequestUTL:function(bidReq){
```
Comment (shlomisut) [inline modules/timBidAdapter.js]: OK, UTL was my typo, but I expect you to notice`URL` is what's needed here

**Comment by borisjaskerovich** [comment]
LGTM

**Comment by borisjaskerovich** [comment]
the response from the server is always the same. i guess this would be resolved.

**Pull request merged**